### PR TITLE
Support 'guard' argument for registration (fixes #39)

### LINKED ISSUE #39: Support passing callable to register which allow to dynamically enable/disable the registration
We have a package that registers for many (different) gossip hooks.
Our package has a "disabling" context. We need a way to disable all our registrations during the context and re-enable them afterward.

### PR DESCRIPTION
**Pull request by ayalash**

### CODE REVIEW

File: doc/advanced.rst [hunk 1]
```diff
@@ -183,6 +183,23 @@ Handlers can be registered with *tokens*. A token is anything that supports equa
 
 
 
+Registration Guard
+------------------
+Sometimes registration should have a way to enable/disable it dynamically. Handlers can be registered with *guard*. A guard is callable which return if the handler should be called.
```
Comment (vmalloc) [inline doc/advanced.rst]: Sometimes it is desirable to have a way to enable/disable registrations dynamically. For this purpose, gossip supports *guards* on handlers. A guard is a callable object, returning whether or not the handler should be triggered

File: doc/changelog.rst [hunk 2]
```diff
@@ -1,6 +1,7 @@
 Changelog
 =========
 
+* :feature:`39` Add ``guard`` arugument to ``register``
```
Comment (vmalloc) [inline doc/changelog.rst]: argument

File: gossip/hooks.py [hunk 3]
```diff
@@ -343,7 +344,8 @@ def define(hook_name, **kwargs):
     return returned
 
 
-def register(func=None, hook_name=None, token=None, tags=None, needs=None, provides=None, reentrant=True, toggles_on=None, toggles_off=None, priority=0):
+def register(func=None, hook_name=None, token=None, tags=None, needs=None, provides=None, reentrant=True,
```
Comment (vmalloc) [inline gossip/hooks.py]: When we drop 2.x support, one of the first things we need to add is keyword-only to this mess :-)

File: doc/changelog.rst (hunk 2 above)
Comment (ayalash) [reply doc/changelog.rst]: Oops... sorry

**Comment by ayalash** [comment]
@vmalloc Fixed. Please review

File: gossip/registration.py [hunk 4]
```diff
@@ -13,10 +13,12 @@
 
 class Registration(object):
 
-    def __init__(self, func, hook, token=None, tags=None, needs=None, provides=None, reentrant=True, toggles_on=None, toggles_off=None, priority=0):
+    def __init__(self, func, hook, token=None, tags=None, needs=None, provides=None, reentrant=True, toggles_on=None,
+                 toggles_off=None, priority=0, guard=None):
         super(Registration, self).__init__()
 
         assert not (toggles_off is not None and toggles_on is not None), 'Cannot specify both toggles_on and toggles_off'
+        assert (guard is None) or callable(guard), "Reistration 'guard' argument must be callable if specified"
```
Comment (vmalloc) [inline gossip/registration.py]: `Registration`

Comment (ayalash) [reply gossip/registration.py]: I should learn how to type :(

**Comment by ayalash** [comment]
@vmalloc Fixed, can you review?
(I hope I don't have anymore typos)

**Pull request merged**